designs/v1-api.md

```diff
@@ -0,0 +1,529 @@
+
+#### Removing `resource.requests` from NodePool Schema
+
+**Category:** Stability
+
+`spec.resource.requests` is part of the NodeClaim API and is set by the Karpenter controller when creating new NodeClaim resources from the scheduling loop. This field describes the required resource requests cacluated from the summation of pod resource requests needed to run against a new instance that we launch. This field is also used to establish Karpenter initialization, where we rely on the presence of requested resources in the newly created Node before we will allow the Karpenter disruption controller to disrupt the node.
+
+We are currently templating every field from the NodeClaim API onto the NodePool `spec.template` (in the same way that a Deployment templates every field from Pod). Validation currently blocks users from setting resource requests in the NodePool, since the field does not make any sense in that context.
+
+Since this field is currently dead API in the schema, we should drop this field from the NodePool at v1.
+
+## NodeClaim API
+
+```
+apiVersion: karpenter.sh/v1
+kind: NodeClaim
+metadata:
+  name: default
+spec:
+  nodeClassRef: 
+    group: karpenter.k8s.aws # Updated since only a single version will be served
+    kind: EC2NodeClass
+    name: default
+  taints:
+    - key: example.com/special-taint
+      effect: NoSchedule
+  startupTaints:
+    - key: example.com/another-taint
+      effect: NoSchedule
+  requirements:
+    - key: "karpenter.k8s.aws/instance-category"
+      operator: In
+      values: ["c", "m", "r"]
+      minValues: 2
+  resources: 
+    requests:
+      cpu: "1"
+      memory: "100Mi"
+status:
+  allocatable:
+    cpu: 1930m
+    ephemeral-storage: 17Gi
+    memory: 3055Mi
+    pods: "29"
+    vpc.amazonaws.com/pod-eni: "9"
+  capacity:
+    cpu: "2"
+    ephemeral-storage: 20Gi
+    memory: 3729Mi
+    pods: "29"
+    vpc.amazonaws.com/pod-eni: "9"
+  conditions:
+    - lastTransitionTime: "2024-02-02T19:54:34Z"
+      status: "True"
+      type: Launched
+    - lastTransitionTime: "2024-02-02T19:54:34Z"
+      status: "True"
+      type: Registered
+    - lastTransitionTime: "2024-02-02T19:54:34Z"
+      status: "True"
+      type: Initialized
+    - lastTransitionTime: "2024-02-02T19:54:34Z"
+      status: "True"
+      type: Drifted
+      reason: RequirementsDrifted
+    - lastTransitionTime: "2024-02-02T19:54:34Z"
+      status: "True"
+      type: Expired
+    - lastTransitionTime: "2024-02-02T19:54:34Z"
+      status: "True"
+      type: Empty
+    - lastTransitionTime: "2024-02-02T19:54:34Z"
+      status: "True"
+      type: Ready
+  nodeName: ip-192-168-71-87.us-west-2.compute.internal
+  providerID: aws:///us-west-2b/i-053c6b324e29d2275
+  imageID: ami-0b1e393fbe12f411c
```

Comment:
How will we handle:
- providers where the machine image specification is a tuple (eg `{image: "e911d3ce-5a58-4e5f-bab1-e5c5492779df", version: 42}`
- node provisioning where there's no image (for example: there's a pool of provisioned bare metal and someone wants Karpenter to wake up servers when needed)

?

It'd be a shame to need a v2 API for those cases.

Reply:
I'm struggling to see a case where an image doesn't exist assigned to a NodeClaim/Instance. In the pre-provisioned case, the NodeClaim will eventually have an image assigned to it when it wakes up. In the tuple case, you could construct a string that would make sense in the context of "imageID" e.g. "e911d3ce-5a58-4e5f-bab1-e5c5492779df/v42"